File: game-engine/src/main/java/org/luxons/sevenwonders/game/data/definitions/WonderSidePickMethod.java
package org.luxons.sevenwonders.game.data.definitions;

import java.util.Random;

public enum WonderSidePickMethod {
    ALL_A {
        @Override
        public WonderSide pickSide(Random random, WonderSide lastPickedSide) {
            return WonderSide.A;
        }
    },
    ALL_B {
        @Override
        public WonderSide pickSide(Random random, WonderSide lastPickedSide) {
            return WonderSide.B;
        }
    },
    EACH_RANDOM {
        @Override
        public WonderSide pickSide(Random random, WonderSide lastPickedSide) {
            return random.nextBoolean() ? WonderSide.A : WonderSide.B;
        }
    },
    SAME_RANDOM_FOR_ALL {
        @Override
        public WonderSide pickSide(Random random, WonderSide lastPickedSide) {
            if (lastPickedSide == null) {
                return random.nextBoolean() ? WonderSide.A : WonderSide.B;
            } else {
                return lastPickedSide;
            }
        }
    };

    public abstract WonderSide pickSide(Random random, WonderSide lastPickedSide);
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/data/WonderSidePickMethod.kt
package org.luxons.sevenwonders.game.data

import java.util.Random

enum class WonderSidePickMethod {
    ALL_A {
        override fun pickSide(random: Random, lastPickedSide: WonderSide?): WonderSide {
            return WonderSide.A
        }
    },
    ALL_B {
        override fun pickSide(random: Random, lastPickedSide: WonderSide?): WonderSide {
            return WonderSide.B
        }
    },
    EACH_RANDOM {
        override fun pickSide(random: Random, lastPickedSide: WonderSide?): WonderSide {
            return if (random.nextBoolean()) WonderSide.A else WonderSide.B
        }
    },
    SAME_RANDOM_FOR_ALL {
        override fun pickSide(random: Random, lastPickedSide: WonderSide?): WonderSide {
            return lastPickedSide ?: if (random.nextBoolean()) WonderSide.A else WonderSide.B
        }
    };

    abstract fun pickSide(random: Random, lastPickedSide: WonderSide?): WonderSide
}
